### PR TITLE
Cra test gql to compass

### PR DESCRIPTION
**Pull request by koala7659**
This PR contains the implementation of GraphQL client used by compass-runtime-agent component tests to connect to Compass Director API and perform various operations needed for testing scenarios:

- Registering/Unregistering Runtime
- Registering/Unregistering application (a remote system)
- Registering/Unregistering Formation (scenario)
- Assign Runtime and Application to Formation (scenario)
- Getting one time token to connect compass-runtime-agent with compass

There is a hidden assumption here: the testing scenario always exists, and we are not adding/removing it from compass. It might be problematic

### CODE REVIEW

File: tests/components/application-connector/Makefile [hunk 1]
```diff
@@ -5,7 +5,7 @@ MOCK_APP_IMAGE = "$(DOCKER_PUSH_REPOSITORY)$(DOCKER_PUSH_DIRECTORY)/mock-app:$(D
 
 .PHONY: release image
 
-release: publish-gateway-test publish-mock-app publish-validator-test publish-compass-runtime-agent-test
+release: publish-compass-runtime-agent-test
```
Comment (franpog859) [inline tests/components/application-connector/Makefile:8]: I think we should revert this change

File: tests/components/application-connector/resources/charts/compass-runtime-agent-test/values.yaml [hunk 2]
```diff
@@ -7,13 +7,11 @@ skipDirectorCertVerification: true
 serviceAccountName: "test-compass-runtime-agent"
 
 containerRegistry:
-  path: "eu.gcr.io/kyma-project"
-
+  path: "docker.io/koala7659"
```
Comment (franpog859) [inline tests/components/application-connector/resources/charts/compass-runtime-agent-test/values.yaml:10]: I think we should revert this change too

File: tests/components/application-connector/resources/charts/compass-runtime-agent-test/values.yaml [hunk 3]
```diff
@@ -7,13 +7,11 @@ skipDirectorCertVerification: true
...
 images:
   compassTest:
     name: "compass-runtime-agent-test"
-    version: "PR-15380"
-
+    version: "PR-1234"
```
Comment (franpog859) [inline tests/components/application-connector/resources/charts/compass-runtime-agent-test/values.yaml:14]: The correct version is `PR-15119`

File: tests/components/application-connector/test/compass-runtime-agent/suite_test.go [hunk 4]
```diff
@@ -1,15 +1,21 @@
...
+	"github.com/kyma-project/kyma/tests/components/application-connector/test/compass-runtime-agent/testkit/graphql"
+	"github.com/kyma-project/kyma/tests/components/application-connector/test/compass-runtime-agent/testkit/oauth"
+	"github.com/pkg/errors"
 	"github.com/stretchr/testify/suite"
+	"github.com/vrischmann/envconfig"
 	"k8s.io/client-go/kubernetes"
 	"k8s.io/client-go/rest"
+	"net/http"
+	"testing"
+	"time"
```
Comment (franpog859) [inline tests/components/application-connector/test/compass-runtime-agent/suite_test.go:18]: Is it correctly formatted?

File: tests/components/application-connector/test/compass-runtime-agent/suite_test.go [hunk 5]
```diff
@@ -44,7 +56,41 @@ func (gs *CompassRuntimeAgentSuite) SetupSuite() {
 
 	applicationGetter := gs.applicationsClientSet.ApplicationconnectorV1alpha1().Applications()
 	gs.appComparator, err = applications.NewComparator(gs.Require(), secretComparator, applicationGetter, "", "")
-	gs.Require().Nil(err)
+	//gs.T().Log("Attempt to unregister application in compass")
```
Comment (franpog859) [inline tests/components/application-connector/test/compass-runtime-agent/suite_test.go:59]: It think we can remove the commented code

File: tests/components/application-connector/test/compass-runtime-agent/testkit/third_party/machinebox/graphql/LICENSE [hunk 6]
```diff
@@ -0,0 +1,201 @@
+                                 Apache License
```
Comment (franpog859) [inline tests/components/application-connector/test/compass-runtime-agent/testkit/third_party/machinebox/graphql/LICENSE:1]: Do we need to use some custom version of this `machinebox/graphql` or we can just use it as a normal dependency?

**Review by franpog859 — COMMENTED** [review]
I'm posting these initial few comments just to speed up the process a little bit. Code review still in progress

File: tests/components/application-connector/test/compass-runtime-agent/config.go [hunk 7]
```diff
@@ -0,0 +1,16 @@
...
+
+import "fmt"
+
+type config struct {
+	DirectorURL                  string `envconfig:"default=http://compass-director.compass-system.svc.cluster.local:3000/graphql"`
+	SkipDirectorCertVerification bool   `envconfig:"default=false"`
+	OauthCredentialsNamespace    string `envconfig:"default=test"`
+	OauthCredentialsSecretName   string `envconfig:"default=oauth-compass-credentials"`
+	TestingTenant                string `envconfig:"default=tenant"`
+}
```
Comment (franpog859) [inline tests/components/application-connector/test/compass-runtime-agent/config.go:11]: The env variables with the Compass data are specified in [this comment](https://github.com/kyma-project/test-infra/pull/5977#issuecomment-1240414698). Do they somehow correspond to this config or these are two separate things?

File: tests/components/application-connector/test/compass-runtime-agent/synchronisation_test.go [hunk 8]
```diff
@@ -19,33 +19,34 @@ type ApplicationReader interface {
...
 func (gs *CompassRuntimeAgentSuite) createAppAndWaitForSync(appReader ApplicationReader, compassAppName, expectedAppName string) (string, error) {
 
-	var runtimeID string
+	var applicationID string
+
+	scenarioName := "auto-testing"
 
 	exec := func() error {
-		id, err := gs.directorClient.RegisterApplication(compassAppName)
+		id, err := gs.directorClient.RegisterApplication(compassAppName, scenarioName)
```
Comment (franpog859) [inline tests/components/application-connector/test/compass-runtime-agent/synchronisation_test.go:47]: Shouldn't it also be assigned to the correct formation to appear in the runtime? I didn't check it manually yet, just reading the code

File: tests/components/application-connector/test/compass-runtime-agent/testkit/director/queryprovider.go [hunk 9]
```diff
@@ -0,0 +1,69 @@
...
+func (qp queryProvider) assignFormationForAppMutation(applicationId, formationName string) string {
+	return fmt.Sprintf(`mutation {
+	result: assignFormation(
+		objectID: "%s"
+		objectType: APPLICATION
+		formation: { name: "%s" }
+	) { id } }`, applicationId, formationName)
+}
+
+func (qp queryProvider) assignFormationForRuntimeMutation(runtimeId, formationName string) string {
```
Comment (franpog859) [inline tests/components/application-connector/test/compass-runtime-agent/testkit/director/queryprovider.go:34]: Shouldn't we also have some mutations to unassign the formations from Applications and Runtimes?

File: tests/components/application-connector/test/compass-runtime-agent/testkit/oauth/types.go [hunk 10]
```diff
@@ -0,0 +1,38 @@
...
+
+const (
+	contentTypeHeader                = "Content-Type"
+	contentTypeApplicationURLEncoded = "application/x-www-form-urlencoded"
+
+	grantTypeFieldName   = "grant_type"
+	credentialsGrantType = "client_credentials"
+
+	scopeFieldName = "scope"
+	scopes         = "application:read application:write formation:write runtime:read runtime:write"
```
Comment (franpog859) [inline tests/components/application-connector/test/compass-runtime-agent/testkit/oauth/types.go:13]: Should we have some `formation` scope?

**Review by franpog859 — COMMENTED** [review]
Some additional comments from the code review. I need to check it manually though

File: tests/components/application-connector/resources/charts/compass-runtime-agent-test/values.yaml (hunk 3 above)
Comment (koala7659) [reply tests/components/application-connector/resources/charts/compass-runtime-agent-test/values.yaml:14]: Fixed

File: tests/components/application-connector/test/compass-runtime-agent/suite_test.go (hunk 5 above)
Comment (koala7659) [reply tests/components/application-connector/test/compass-runtime-agent/suite_test.go:59]: Removed

File: tests/components/application-connector/test/compass-runtime-agent/suite_test.go (hunk 4 above)
Comment (koala7659) [reply tests/components/application-connector/test/compass-runtime-agent/suite_test.go:18]: Fixed

File: tests/components/application-connector/test/compass-runtime-agent/synchronisation_test.go (hunk 8 above)
Comment (koala7659) [reply tests/components/application-connector/test/compass-runtime-agent/synchronisation_test.go:47]: I'm not sure since I did not write this particular piece. We will see during tests

File: tests/components/application-connector/test/compass-runtime-agent/suite_test.go (hunk 4 above)
Comment (koala7659) [reply tests/components/application-connector/test/compass-runtime-agent/suite_test.go:18]: I run go fmt and there were no changes

File: tests/components/application-connector/test/compass-runtime-agent/config.go (hunk 7 above)
Comment (koala7659) [reply tests/components/application-connector/test/compass-runtime-agent/config.go:11]: All needed env variables will be provided via in secret named  `OauthCredentialsSecretName`

File: tests/components/application-connector/test/compass-runtime-agent/testkit/oauth/types.go (hunk 10 above)
Comment (koala7659) [reply tests/components/application-connector/test/compass-runtime-agent/testkit/oauth/types.go:13]: We have `formation:write` and it seems to work fine.

File: tests/components/application-connector/test/compass-runtime-agent/testkit/third_party/machinebox/graphql/LICENSE (hunk 6 above)
Comment (koala7659) [reply tests/components/application-connector/test/compass-runtime-agent/testkit/third_party/machinebox/graphql/LICENSE:1]: There was a reason for that but I don't remember it :(

File: tests/components/application-connector/resources/charts/compass-runtime-agent-test/values.yaml (hunk 2 above)
Comment (koala7659) [reply tests/components/application-connector/resources/charts/compass-runtime-agent-test/values.yaml]: Done

File: tests/components/application-connector/test/compass-runtime-agent/testkit/director/queryprovider.go (hunk 9 above)
Comment (koala7659) [reply tests/components/application-connector/test/compass-runtime-agent/testkit/director/queryprovider.go:34]: During cleanup it is not necessary. We can just delete Runtime and application. But I think I forgot one thing adding /removing formation. We should discuss it

File: tests/components/application-connector/test/compass-runtime-agent/testkit/third_party/machinebox/graphql/LICENSE (hunk 6 above)
Comment (franpog859) [reply tests/components/application-connector/test/compass-runtime-agent/testkit/third_party/machinebox/graphql/LICENSE:1]: Ok

File: tests/components/application-connector/test/compass-runtime-agent/synchronisation_test.go (hunk 8 above)
Comment (franpog859) [reply tests/components/application-connector/test/compass-runtime-agent/synchronisation_test.go:47]: Ok

File: tests/components/application-connector/Makefile (hunk 1 above)
Comment (koala7659) [reply tests/components/application-connector/Makefile]: Fixed

**Review by franpog859 — APPROVED** [review]
The code looks good

**Review by majakurcius — APPROVED** [review]
